### PR TITLE
Jetson Update for CI Redo

### PR DESCRIPTION
**Pull request by puririshi98**
cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

**Review by ezyang — APPROVED** [review]

**Review by malfet — CHANGES_REQUESTED** [review]
Name feels really confusing.
`HAS_TRITION` feels like a good flag, or `IS_ARM_LINUX`
But `IS_JETSON` is really confusing, isn't it.
Also, please file an issue that tracks enablement of all those tests or explaining why they must be skipped on ARM+CUDA platform forever

File: test/inductor/test_minifier.py [hunk 1]
```diff
@@ -149,6 +151,7 @@ def inner(x):
         self.assertEqual(test_proc.returncode, repro_proc.returncode)
         self.assertNotEqual(test_proc.returncode, 0)
 
+    @unittest.skipIf(not _HAS_TRITON, "openAI Triton unavailable")
     def test_after_aot_cpu_runtime_error(self):
         self._test_after_aot_runtime_error("cpu", CPP_RUNTIME_ERROR)
```
Comment (peterbell10) [inline test/inductor/test_minifier.py:156]: Why do these cpu tests require triton?

Comment (puririshi98) [reply test/inductor/test_minifier.py:156]: w/o this flag we get the following errors:
```
FAIL: test_after_aot_cpu_accuracy_backend_passes (__main__.MinifierTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "inductor/test_minifier.py", line 191, in test_after_aot_cpu_accuracy_backend_passes
    self._test_after_aot_backend_passes("cpu", 4, CPP_ACCURACY_ERROR)
  File "inductor/test_minifier.py", line 181, in _test_after_aot_backend_passes
    self.assertEqual(proc.returncode, 0)
  File "/home/riship/.local/lib/python3.8/site-packages/torch/testing/_internal/common_utils.py", line 2859, in assertEqual
    assert_equal(
  File "/home/riship/.local/lib/python3.8/site-packages/torch/testing/_comparison.py", line 1270, in assert_equal
    raise error_metas[0].to_error(msg)
AssertionError: Scalars are not equal!

Absolute difference: 1
Relative difference: inf

======================================================================
FAIL: test_after_aot_cpu_accuracy_error (__main__.MinifierTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "inductor/test_minifier.py", line 108, in test_after_aot_cpu_accuracy_error
    (tb1, tb2), _ = self._test_after_aot("cpu", CPP_ACCURACY_ERROR, 4)
  File "inductor/test_minifier.py", line 94, in _test_after_aot
    (test_proc, _, repro_proc), _ = self._run_full_test(
  File "/home/riship/.local/lib/python3.8/site-packages/torch/_dynamo/test_minifier_common.py", line 126, in _run_full_test
    self.assertIsNotNone(repro_dir)
AssertionError: unexpectedly None

======================================================================
FAIL: test_after_aot_cpu_compile_backend_passes (__main__.MinifierTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "inductor/test_minifier.py", line 185, in test_after_aot_cpu_compile_backend_passes
    self._test_after_aot_backend_passes("cpu", 2, CPP_COMPILE_ERROR)
  File "inductor/test_minifier.py", line 181, in _test_after_aot_backend_passes
    self.assertEqual(proc.returncode, 0)
  File "/home/riship/.local/lib/python3.8/site-packages/torch/testing/_internal/common_utils.py", line 2859, in assertEqual
    assert_equal(
  File "/home/riship/.local/lib/python3.8/site-packages/torch/testing/_comparison.py", line 1270, in assert_equal
    raise error_metas[0].to_error(msg)
AssertionError: Scalars are not equal!

Absolute difference: 1
Relative difference: inf

======================================================================
FAIL: test_after_aot_cpu_compile_error (__main__.MinifierTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "inductor/test_minifier.py", line 103, in test_after_aot_cpu_compile_error
    (tb1, tb2), _ = self._test_after_aot("cpu", CPP_COMPILE_ERROR, 2)
  File "inductor/test_minifier.py", line 94, in _test_after_aot
    (test_proc, _, repro_proc), _ = self._run_full_test(
  File "/home/riship/.local/lib/python3.8/site-packages/torch/_dynamo/test_minifier_common.py", line 126, in _run_full_test
    self.assertIsNotNone(repro_dir)
AssertionError: unexpectedly None
======================================================================
FAIL: test_after_aot_cpu_runtime_backend_passes (__main__.MinifierTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "inductor/test_minifier.py", line 188, in test_after_aot_cpu_runtime_backend_passes
    self._test_after_aot_backend_passes("cpu", 2, CPP_RUNTIME_ERROR)
  File "inductor/test_minifier.py", line 181, in _test_after_aot_backend_passes
    self.assertEqual(proc.returncode, 0)
  File "/home/riship/.local/lib/python3.8/site-packages/torch/testing/_internal/common_utils.py", line 2859, in assertEqual
    assert_equal(
  File "/home/riship/.local/lib/python3.8/site-packages/torch/testing/_comparison.py", line 1270, in assert_equal
    raise error_metas[0].to_error(msg)
AssertionError: Scalars are not equal!

Absolute difference: 1
Relative difference: inf

======================================================================
FAIL: test_after_aot_cpu_runtime_error (__main__.MinifierTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "inductor/test_minifier.py", line 153, in test_after_aot_cpu_runtime_error
    self._test_after_aot_runtime_error("cpu", CPP_RUNTIME_ERROR)
  File "inductor/test_minifier.py", line 143, in _test_after_aot_runtime_error
    (test_proc, _, repro_proc), _ = self._run_full_test(
  File "/home/riship/.local/lib/python3.8/site-packages/torch/_dynamo/test_minifier_common.py", line 126, in _run_full_test
    self.assertIsNotNone(repro_dir)
AssertionError: unexpectedly None

----------------------------------------------------------------------
Ran 12 tests in 37.946s

FAILED (failures=6, skipped=6)
```

**Comment by puririshi98** [comment]
> Name feels really confusing. `HAS_TRITION` feels like a good flag, or `IS_ARM_LINUX` But `IS_JETSON` is really confusing, isn't it. Also, please file an issue that tracks enablement of all those tests or explaining why they must be skipped on ARM+CUDA platform forever

IS_JETSON makes the same sense as IS_ARM_LINUX to me. It means the current platform "is jetson". This is the exact same meaning as IS_ARM_LINUX, which implies the current platofrm "is arm linux".  I see a direct 1:1 comparison between the two naming conventions. Regardless here is the issue: https://github.com/pytorch/pytorch/issues/94614

**Comment by puririshi98** [comment]
> Name feels really confusing. HAS_TRITION feels like a good flag, or IS_ARM_LINUX But IS_JETSON is really confusing, isn't it. Also, please file an issue that tracks enablement of all those tests or explaining why they must be skipped on ARM+CUDA platform forever
@malfet 
IS_JETSON makes the same sense as IS_ARM_LINUX to me. It means the current platform "is jetson". This is the exact same meaning as IS_ARM_LINUX, which implies the current platofrm "is arm linux". I see a direct 1:1 comparison between the two naming conventions. Regardless here is the issue: https://github.com/pytorch/pytorch/issues/94614

File: test/inductor/test_minifier.py (hunk 1 above)
Comment (peterbell10) [reply test/inductor/test_minifier.py:156]: If it was requiring triton I would expect an import error. You should just skip on jetson.

**Comment by puririshi98** [comment]
@malfet @peterbell10 anything left needed to merge?

**Comment by puririshi98** [comment]
@cpuhrsch is there anything else needed to merge?

**Comment by puririshi98** [comment]
@cpuhrsch @ezyang @malfet @peterbell10, are there any additional updates I should make to get this merged in?

**Comment by ezyang** [comment]
Was waiting on @malfet

**Comment by malfet** [comment]
> @malfet IS_JETSON makes the same sense as IS_ARM_LINUX to me. 

Not sure I understand what do you mean by this one.
My point, that `IS_ARM_LINUX` guard would make sure that test would be skipped on all Arm powered linux platforms, regardless of the vendor, wheres `IS_JETSON` will skip it only for NVIDIA's ARM chips. I'm not sure why one needs to make a distinction here? Also, Arm([wiki article](https://en.wikipedia.org/wiki/Arm_(disambiguation))), as CPU architecture, is more easily recognizable by people reading the code than Jetson([wiki article](https://en.wikipedia.org/wiki/Jetson))

**Comment by puririshi98** [comment]
the difference is that IS_ARM_LINUX covers all arm linux platforms but jetsons are special. they combine nvidia gpu and arm cpu ram in a complicated way. this leads to failures that do not occur on all arm linux servers, just jetson. so IS_JETSON is the right naming convention. cc @ptrblck

File: test/inductor/test_smoke.py [hunk 2]
```diff
@@ -24,6 +25,7 @@ def _test_f(x):
 
 
 class SmokeTest(TestCase):
+    @unittest.skipIf(IS_JETSON, "openAI Triton not available for Jetson")
```
Comment (malfet) [inline test/inductor/test_smoke.py:28]: This should just skip based on  (which just checks whether triton is installed or not)
https://github.com/pytorch/pytorch/blob/641cb4243c2ed54b02e713d03659824034b6c82f/torch/testing/_internal/inductor_utils.py#L23
```suggestion
    @unittest.skipIf(torch._internal.inductor_utils.HAS_CUDA, "Triton is not available")
```

Comment (puririshi98) [reply test/inductor/test_smoke.py]: https://github.com/pytorch/pytorch/actions/runs/4197589008/jobs/7280465927
your suggested change is failing CI. should i revert to my original IS_JETSON skipIf or is there some quick fix for the failure ?

File: test/test_torch.py [hunk 3]
```diff
@@ -1979,13 +1979,15 @@ def _generate_correlation_tensors(self, device, dtype):
...
+    @skipDtypeForJetsonCPU(torch.cfloat)
     def test_corrcoef(self, device, dtype):
         for x in self._generate_correlation_tensors(device, dtype):
             res = torch.corrcoef(x)
             ref = np.corrcoef(x.cpu().numpy())
             self.assertEqual(res, ref, exact_dtype=False)
 
     @dtypes(torch.int, torch.float, torch.cfloat)
+    @skipDtypeForJetsonCPU(torch.cfloat)
     def test_cov(self, device, dtype):
```
Comment (malfet) [inline test/test_torch.py:1991]: Hmm, what errors are you see when running following command on Jetson?
```
python test_torch.py -v -k test_corrcoef_cpu_complex64
```

Comment (puririshi98) [reply test/test_torch.py]: yes this used to be needed but no longer, thanks for pointing it out

File: test/test_public_bindings.py [hunk 4]
```diff
@@ -271,7 +271,7 @@ def test_no_new_bindings(self):
         self.assertTrue(torch_C_bindings.issubset(torch_C_allowlist_superset), msg)
 
     # AttributeError: module 'torch.distributed' has no attribute '_shard'
-    @unittest.skipIf(IS_WINDOWS, "Distributed Attribute Error")
+    @unittest.skipIf(IS_WINDOWS or IS_JETSON, "Distributed Attribute Error")
```
Comment (malfet) [inline test/test_public_bindings.py:274]: This feels like a config problem to me (i.e. you compile pytorch without some deps, but whatever)

File: test/inductor/test_smoke.py [hunk 5]
```diff
@@ -1,10 +1,11 @@
 # Owner(s): ["module: inductor"]
 import logging
+import unittest
 
 import torch
 import torch._dynamo as torchdynamo
 import torch._inductor.config as torchinductor_config
-from torch.testing._internal.common_utils import IS_LINUX, TestCase
+from torch.testing._internal.common_utils import IS_JETSON, IS_LINUX, TestCase
```
Comment (malfet) [inline test/inductor/test_smoke.py:8]: You don't need is jetson here I guess, but would be nice to add
```suggestion
from torch.testing._internal.common_utils import IS_LINUX, TestCase
from torch.testing._internal.inductor_utils import HAS_CUDA
```

File: test/nn/test_embedding.py [hunk 6]
```diff
@@ -1170,6 +1170,8 @@ def _test_EmbeddingBag(
     @dtypesIfCUDA(*itertools.product((torch.int, torch.long), (torch.int, torch.long),
                                      (torch.float, torch.double, torch.half)))
     def test_embedding_bag_device(self, device, dtypes):
+        if torch.bfloat16 in dtypes and device=="cpu":
+            self.skipTest("bfloat16 not supported with Jetson cpu")
```
Comment (malfet) [inline test/nn/test_embedding.py:1174]: Err? This would disable the test on all CPU, not just on jetson?

Comment (puririshi98) [reply test/nn/test_embedding.py]: sorry typo from recent commit

**Comment by malfet** [comment]
> the difference is that IS_ARM_LINUX covers all arm linux platforms but jetsons are special. they combine nvidia gpu and arm cpu ram in a complicated way. this leads to failures that do not occur on all arm linux servers, just jetson. so IS_JETSON is the right naming convention. cc @ptrblck

I completely understand the argument about CPU+GPU caches synchronization, so have no objection to CUDA specific changes, but it should not affect CPU-only part of tests, where Jetson is just another ARM chip, isn't it?

**Review by malfet — APPROVED** [review]
Some of the exclusions are weird, but sure, LGTM

File: test/test_public_bindings.py (hunk 4 above)
Comment (puririshi98) [reply test/test_public_bindings.py:274]: yes pytorch is built w/o distributed for jetsons

**Comment by puririshi98** [comment]
> > the difference is that IS_ARM_LINUX covers all arm linux platforms but jetsons are special. they combine nvidia gpu and arm cpu ram in a complicated way. this leads to failures that do not occur on all arm linux servers, just jetson. so IS_JETSON is the right naming convention. cc @ptrblck
> 
> I completely understand the argument about CPU+GPU caches synchronization, so have no objection to CUDA specific changes, but it should not affect CPU-only part of tests, where Jetson is just another ARM chip, isn't it?

I am not sure about the details of the jetson hardware but we have found it to not behave the same as other ARM systems. If any of these issue is found to reproduce in our ARM server CI (which has not been the case up until now) we can replace IS_JETSON for the more general IS_ARM_LINUX. IS_JETSON should suffice for now though

**Comment by puririshi98** [comment]
@cpuhrsch @ezyang @malfet @peterbell10 looks like its ready for merging. could one of you merge when you get a chance? i do not have permission

**Pull request closed without merging**